File: server_integration_tests/java/java_server/src/main/java/dev/graphlink/javaserver/impl/AuthorSchemaMappingsServiceImpl.java
package dev.graphlink.javaserver.impl;

import dev.graphlink.javaserver.generated.interfaces.GlArticleProjection;
import dev.graphlink.javaserver.generated.services.AuthorSchemaMappingsService;
import dev.graphlink.javaserver.generated.types.Article;
import dev.graphlink.javaserver.generated.types.Author;
import graphql.GraphQLContext;

import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class AuthorSchemaMappingsServiceImpl implements AuthorSchemaMappingsService {

    @Override
     public Map<Author, List<GlArticleProjection>> authorArticles(List<Author> value, GraphQLContext graphQLContext) {
        Map<Author, List<GlArticleProjection>> result = new HashMap<>();
        for (Author author : value) {
            List<GlArticleProjection> articles = Data.articles.stream()
                .filter(a -> a.getAuthorId().equals(author.getId()))
                .collect(Collectors.toList());
            result.put(author, articles);
        }
        return result;
    }

    @Override
    public List<Article> authorLatestArticles(int limit, Author value) {
        return Data.articles.stream()
            .filter(a -> a.getAuthorId().equals(value.getId()))
            .limit(limit)
            .collect(Collectors.toList());
    }
}

File: server_integration_tests/kotlin/kotlin_server_blocking/src/main/kotlin/dev/graphlink/kotlinserverblocking/impl/AuthorSchemaMappingsServiceImpl.kt
package dev.graphlink.kotlinserverblocking.impl

import dev.graphlink.kotlinserverblocking.generated.interfaces.Artcile
import dev.graphlink.kotlinserverblocking.generated.services.AuthorSchemaMappingsService
import dev.graphlink.kotlinserverblocking.generated.types.Article
import dev.graphlink.kotlinserverblocking.generated.types.Author
import graphql.GraphQLContext
import org.springframework.stereotype.Service

@Service
class AuthorSchemaMappingsServiceImpl : AuthorSchemaMappingsService {

    override fun authorArticles(
        value: List<Author>,
        graphQLContext: GraphQLContext,
    ): Map<Author, List<Artcile>?> {
        val result = mutableMapOf<Author, List<Artcile>?>()
        for (author in value) {
            result[author] = Data.articles.filter { it.authorId == author.id }
        }
        return result
    }

    override fun authorLatestArticles(limit: Int, value: Author): List<Article> =
        Data.articles.filter { it.authorId == value.id }.take(limit)
}
